lib/Performance/Probability.pm

```diff
@@ -0,0 +1,233 @@
+}
+
+#Sigma( (profit**2) * winning probability + (loss**2) * losing probability ).
+sub _variance_x_square {
+
+    my $pk = shift;
+    my $lk = shift;
+    my $wk = shift;
+
+    my $sum = 0;
+    my $i;
+
+    for ($i = 0; $i < @{$wk}; ++$i) {
+        $sum = $sum + (($wk->[$i]**2) * $pk->[$i]) + (($lk->[$i]**2) * (1 - $pk->[$i]));
+    }
+
+    return $sum;
+}
+
+#Sum of Covariance(i,j). See the documentation for the details.
+#Covariance(i, j) is the covariance between contract i and j with time overlap.
+sub _covariance {
+
+    my ($start_time, $sell_time, $underlying, $types, $pk, $lk, $wk) = @_;
+
+    my ($i, $j);
+    my $covariance = 0;
+
+    for ($i = 0; $i < @{$start_time}; ++$i) {
+        for ($j = 0; $j < @{$sell_time}; ++$j) {
+            if ($i != $j and $underlying->[$i] eq $underlying->[$j]) {
+
+                #check for time overlap.
+                my ($start_i, $start_j, $sell_i, $sell_j);
+                $start_i = $start_time->[$i];
+                $start_j = $start_time->[$j];
+                $sell_i  = $sell_time->[$i];
+                $sell_j  = $sell_time->[$j];
+
+                if ($start_j > $start_i and $start_j < $sell_i) {
+                    #calculate a, b and c interval. please see the documentation for details
+                    #of a, b and c.
+                    my $a_interval = $start_j - $start_i;
+                    my $b_interval = $sell_i - $start_j;
+                    my $c_interval = $sell_j - $sell_i;
+
+                    if ($c_interval < 0) {
+                        $c_interval = 0 - $c_interval;
+                        $b_interval = $sell_j - $start_i;
+                    }
+
+                    my $i_strike = 0.0 - Math::Gauss::XS::inv_cdf($pk->[$i]);
+                    my $j_strike = 0.0 - Math::Gauss::XS::inv_cdf($pk->[$j]);
+
+                    my $corr_ij = $b_interval / (sqrt($a_interval + $b_interval) * sqrt($b_interval + $c_interval));
+
+                    if ($types->[$i] ne $types->[$j]) {
+                        $corr_ij = -1 * $corr_ij;
+                    }
+
+                    my $p_ij = Math::BivariateCDF::bivnor($i_strike, $j_strike, $corr_ij);
+
+                    my $covariance_ij =
+                        ($p_ij - $pk->[$i] * $pk->[$j]) * ($wk->[$i] - $lk->[$i]) * ($wk->[$j] - $lk->[$j]);
+
+                    $covariance = $covariance + $covariance_ij;
+                }
+            }
+        }
+    }
+
+    return $covariance;
+}
+
+=item B<get_performance_probability>
+
+Calculate performance probability ( modified sharpe ratio );
+
+=cut
+
+sub get_performance_probability {
+
+    my $params = shift;
+
+    my $pnl = $params->{pnl};
+
+    if (not defined $pnl) {
+        die "pnl is a required parameter.";
+    }
+
+    #Below variables are all arrays.
+    my $start_time   = $params->{start_time};
+    my $sell_time    = $params->{sell_time};
+    my $types        = $params->{types};
+    my $underlying   = $params->{underlying};
```

Comment:
Since this is a CPAN module, we should not think about our code base here. I prefer to call this underlying or underlying_symbol which is a bit longer.